### ISSUE TITLE
Pipeline breaks with too large metagenome after assembly (here: 122 million contigs)

### ISSUE THREAD

**Issue by d4straub** [nf-core/mag#267]
## Description of the bug

With _very_ large metagenomes (here: 122 million contigs), process `NFCORE_MAG:MAG:METABAT2_BINNING:BOWTIE2_ASSEMBLY_BUILD` fails with 
```
Command error:
  WARNING: Skipping mount /var/singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container
  [E::bam_hdr_write] Header too long for BAM format
  [main_samview] failed to write the SAM header
  samtools sort: failed to read header from "-"
```

## Steps to reproduce

I guess by taking any fasta file with many sequences, making a bowtie2 index, mapping an arbitraty number of sequencing reads to it, and converting the result to BAM format.
I did test with different samtools versions. The pipeline specific container uses samtools 1.11, I tested with 1.14, same result.

## Cause of the issue:

The header of BAM files is restricted to a [specific size](https://github.com/samtools/samtools/issues/1105#issuecomment-530300080). The SAM header seems to include all reference sequences with name and length information (lines starting with `@SG`). This SAM header seems not size restricted, but when converting the file to BAM, the header may only encompass a specific size.
This limit is also documented in section 4.2 of the [SAM/BAM specification](https://samtools.github.io/hts-specs/SAMv1.pdf).

## Potential fixes:

1. use `--megahit_options` with higher contig length threshold, e.g. `--megahit_options "--min-contig-len 1500"` (default is 200, increase to sensible number)
2. filter SAM so that its header is small enough before converting to BAM
3. decreasing header size by renaming assembled contigs, e.g. with the hexagesimal system, proposed [here](https://github.com/samtools/samtools/issues/633#issuecomment-278378906)
4. filter reference sequences (i.e. reducing metagenome contigs) before making bowtie index and aligning reads

## My fix:

For the metagenome in question I used solution 3: I filtered before indexing and alignment all contigs below 1500bp size from the assembly. I did this with the existing `split_fasta.py` using `split_fasta.py <assembly file> 100000000 1000000000 1500`. I chose the threshold of 1500bp contig length because this step is producing depth information for contig binning with MataBAT2, which is only considering contigs above 1500bp length anyway.
However, I do not think this is optimal, because there might be false positives alignments (i.e. reads aligning ambiguously to >1500bp and <1500bp contig would be labelled now unambiguously aligned).
I think it would be best to use solution 2, but I am not sure whether this is feasible (I did not attempt it).

**Comment by d4straub** [comment]
Thanks for commenting here. I avoided this issue for another recent metagenome assembly with version 3.2.1 by using `--megahit_options "--min-contig-len 500"`.
I am not sure whether thats really an edge case or rather a pretty common one for a medium sized assembly.

Recently we came across a limiting step in another pipeline and the solution seems to simply use chunks via https://nextflow-io.github.io/patterns/process-per-file-chunk/, i.e. there would be several BAM (and could be processed and then merged?). I didnt think that through properly, maybe it has detrimental effects to use a chunked reference or isnt feasible in another way.
Another approach might be to use a different file format, I heard (but never used) of .cram that Rike et al. used in Sarek for speed and size optimization, potentially that has other technical properties. Also no idea if thats feasible in any way, just thoughts if anyone would be to tackle that issue.

**Comment by d4straub** [comment]
Forgot with my most recent assembly to adjust minimal contig length settings and arrived again at that issue. The assembly has 87,633,230 contigs with >=200bp. That is quite annoying indeed.
I thought about the problem a bit more. I would find it ideal to include a step after assembly that checks whether the BAM header is expected to be too large and reduce the assembly to the optimal size (i.e. maximal size that will not fail, omitting the smallest contigs) before `BOWTIE2_ASSEMBLY_BUILD`. I am just not sure how to find the appropriate threshold.

**Comment by jfy133** [comment]
Conceptually I think that makes reasonable sense to me!

We are already now getting the length stats of each bin (`binning.nf`) no reason to not push that back?

**Comment by prototaxites** [comment]
Is there any reason not to just apply a minimum contig length across the board prior to binning? The binning tools all want contigs over (usually) 1 - 1.5 kb anyway, we could enforce that with a parameter. We can publish the raw assembly as usual but pass through the filtered assembly to binning.

If you run into the above problem then we can document that the threshold can just be increased?

**Comment by jfy133** [comment]
Yes exactly that's what I meant by 'pushing back'! Maybe you can add this to your list @d4straub ? 😬

**Comment by d4straub** [comment]
Generally I would like to assemble as much as possible and drop as least as possible data. I agree that bins are most interesting, but the assembly itself can also be very interesting to look at. Bins contain only a relatively small fraction of the assembly, in my experience. So I would like to keep as much of the assembly and data as possible.
I dont know beforehand what contig length threshold will work. So that would be a guessing game. Until now, 500bp or 1500bp reduced the contigs sufficiently to avoid that issue. But the problem is that when that problem occurs, either the pipeline needs to be re-run with a higher min contig length threshold (takes days to weeks of runtime, what a waste of time and power) or a manual invention is needed (my time is wasted). To find the optimal length threshold, not to loose too much sequence info, several re-runs/manual attempts would need to be made. That isnt really optimal in my opinion. 
I would prefer if I can infer from previous information, lets say from the SAM header or the contig headers, how large will be the BAM header and whether that is above the maximum BAM header size. If no, no action required, if yes, smaller to larger contigs could be removed to find the minimal contig threshold that will generate a viable BAM and the assembly could be filtered with that threshold automatically, without manual invention. However, I am not sure how to calculate the SAM/BAM header size or whether thats feasible to implement at all.

**Comment by prototaxites** [comment]
>Generally I would like to assemble as much as possible and drop as least as possible data. I agree that bins are most interesting, but the assembly itself can also be very interesting to look at. Bins contain only a relatively small fraction of the assembly, in my experience. So I would like to keep as much of the assembly and data as possible.

This is why I would only suggest filtering internally for read mapping -> binning - we should not emit filtered raw assemblies. But I take your point about manual fine-tuning!

Would we side-step this problem if we switched to outputting (non-reference-compressed) CRAM instead of BAM? Would also have the advantage of reducing size on disk...

**Comment by d4straub** [comment]
> This is why I would only suggest filtering internally for read mapping -> binning - we should not emit filtered raw assemblies.

Ah sorry I misunderstood. Yes I agree.

> Would we side-step this problem if we switched to outputting (non-reference-compressed) CRAM instead of BAM? Would also have the advantage of reducing size on disk...

Potentially, I dont know if there is any such threshold for CRAM files.
